cmip6_downscaling/methods/regions.py

```diff
@@ -0,0 +1,94 @@
+from typing import Any, Dict, Tuple, Union
+
+import numpy as np
+import regionmask
+import xarray as xr
+
+
+def generate_subdomains(
+    ex_output_grid: Union[xr.Dataset, xr.DataArray],
+    buffer_size: Union[float, int],
+    region_def: str = 'ar6',
+) -> Tuple[Dict[Union[int, float], Any], xr.DataArray]:
+    """
+    Given an example output grid, determine all subdomains that need to be process in order to generate the final output.
+    Outputs the list of bounding boxes for each subdomain considering the buffer size, as well as a mask in the resolution of the example output grid specifying
+    which subdomain's value to use for each grid cell.
+
+    Parameters
+    ----------
+    ex_output_grid : xarray dataarray or dataset. example output grid definition. both the bounding box and resolution in lat/lon directions will be used.
+    buffer_size    : int/float. for each subdomain, how much buffer area to run.
+    region_def     : str. whether to use SREX region definition or ar6 region definition to define subdomain.
+                     see the docs https://regionmask.readthedocs.io/en/stable/defined_scientific.html for more details.
+
+    Returns
+    -------
+    subdomains : dictionary mapping subdomain code to bounding boxes ([min_lon, min_lat, max_lon, max_lat]) for each subdomain
+    mask       : mask of which subdomain code to use for each grid cell
+    """
+    if region_def == 'ar6':
+        regions = regionmask.defined_regions.ar6.land
+    elif region_def == 'srex':
+        regions = regionmask.defined_regions.srex
+    else:
+        raise NotImplementedError('region_def must be eitehr ar6 or srex')
+
+    mask = regions.mask(ex_output_grid)
+    region_codes = np.unique(mask.values)
+    region_codes = region_codes[~np.isnan(region_codes)]
+
+    subdomains = {}
+    for n, bound in zip(regions.numbers, regions.bounds):
+        if n in region_codes:
```

Comment:
this would happen if the `ex_output_grid` doesn't span the globe (e.g. only includes continental us), which in turn allows for testing the workflow with a smaller region

Reply:
oh excellent.